### PR TITLE
Fix #284 crash adding jetpack site

### LINKED ISSUE #284: Crash when adding a self hosted site
Originally from [here](https://github.com/wordpress-mobile/WordPress-Android/pull/4455#issuecomment-242238265). In the Example App:

* Log into .com account with a Jetpack site
* Log into that Jetpack site as self-hosted
* Site -> Fetch All sites
* 💥 

```
Caused by: android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: SiteModel.SITE_ID, SiteModel.URL (code 2067)
    at android.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
    at android.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:732)
    at android.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:754)
    at android.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:64)
    at android.database.sqlite.SQLiteDatabase.updateWithOnConflict(SQLiteDatabase.java:1576)
    at android.database.sqlite.SQLiteDatabase.update(SQLiteDatabase.java:1522)
    at com.yarolegovich.wellsql.UpdateQuery.execute(UpdateQuery.java:79)
    at org.wordpress.android.fluxc.persistence.SiteSqlUtils.insertOrUpdateSite(SiteSqlUtils.java:87)
    at org.wordpress.android.fluxc.store.SiteStore.updateSite(SiteStore.java:566)
    at org.wordpress.android.fluxc.store.SiteStore.onAction(SiteStore.java:502)
    at java.lang.reflect.Method.invoke(Native Method)
    at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
    at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479) 
    at org.greenrobot.eventbus.AsyncPoster.run(AsyncPoster.java:46) 
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
    at java.lang.Thread.run(Thread.java:818) 
```

### PR DESCRIPTION
**Pull request by maxme**
Fix #284 

At least 2 cases to test in the example:

1. Jetpack -> self hosted
    * Clear app data
    * Sign in a .com account with a Jetpack site
    * Try to add that Jetpack site as a self hosted site
    * You should get a "DUPLICATE_SITE" error
1. Self hosted -> Jetpack (with the same account / sites)
    * Clear app data
    * Add that Jetpack site as a self hosted site
    * Sign in a .com account
    * You should get the same number of sites, and the previous self hosted site is now considered a Jetpack

Not easy to test automatically without predefined sites or complex mocks.

I'm not a big fan of [checking for the XMLRPC endpoint ](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/8aad65ddc0fa7d458702792aa0df8e47df9bcc18/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java#L87)as a "key", but I don't think we can find anything better.

### CODE REVIEW

**Comment by aforcier** [comment]
The unit tests need to be updated to deal with `DuplicateSiteException`.

**Comment by maxme** [comment]
@aforcier I updated unit tests and added 2 new tests, also I changed a bit how we generated previous sites (wpcom boolean was off on jetpack sites).

File: example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java [hunk 1]
```diff
@@ -29,7 +29,7 @@ public static SiteModel generateJetpackSite() {
         SiteModel example = new SiteModel();
         example.setSiteId(982);
         example.setSelfHostedSiteId(8);
-        example.setIsWPCom(false);
```
Comment (aforcier) [inline example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java]: I think the intention here originally was that this was generating a _self-hosted_ Jetpack-connected site, so this was accurate (but the method should be renamed for clarity).

Comment (maxme) [inline example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java]: I wasn't sure - but now I'm confused "self-hosted Jetpack-connected site" means "wpcom = true", or maybe you mean "self-hosted Jetpack site connected via XMLRPC" but in that case siteId shouldn't be set.

Comment (aforcier) [inline example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java]: Yes, sorry that's what I meant, connected via XML-RPC.

I thought we did set the `siteId` in those cases: https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/80be01a42eb40a1fc4099251c36b30f341089a2e/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/site/SiteXMLRPCClient.java#L202

Comment (maxme) [inline example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java]: Yes, you're right we set the jetpack_client_id in there. I'll change it back then and rename `generateJetpackSite` to `generateJetpackSiteConnectedViaXMLRPC`

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java [hunk 2]
```diff
@@ -55,19 +60,23 @@ public static int getNumberOfSitesWith(String field, boolean value) {
...
         List<SiteModel> siteResult = WellSql.select(SiteModel.class)
                 .where().beginGroup()
                 .equals(SiteModelTable.ID, site.getId())
                 .endGroup().endWhere().getAsModel();
 
+
         // Looks like a new site, make sure we don't already have it.
         if (siteResult.isEmpty()) {
+            AppLog.d(T.DB, "Site not found using local id");
             // TODO: Make the URL enough, we could get surprise with the site id with .org sites becoming jetpack sites
```
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java]: Can we drop this `TODO` now?

File: example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java (hunk 1 above)
Comment (maxme) [inline example/src/test/java/org/wordpress/android/fluxc/utils/SiteUtils.java]: Fixed in 	b38f78a

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java (hunk 2 above)
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java]: Fixed incomprehensible TODO in 57bf932

File: example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java [hunk 3]
```diff
@@ -272,13 +272,52 @@ public void testGetWPComSites() throws DuplicateSiteException {
...
         List<SiteModel> wpComSites = SiteSqlUtils.getAllWPComSites();
 
-        assertEquals(2, wpComSites.size());
-        for (SiteModel site : wpComSites) {
-            assertNotEquals(jetpackSiteOverSelfHosted.getId(), site.getId());
+        assertEquals(3, wpComSites.size());
+    }
+
+    @Test
+    public void testInsertDuplicateSites() throws DuplicateSiteException {
```
Comment (aforcier) [inline example/src/test/java/org/wordpress/android/fluxc/site/SiteStoreUnitTest.java]: 👍 glad this is here, hard to make an automated test for this case with a live site.

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java (hunk 2 above)
Comment (aforcier) [inline fluxc/src/main/java/org/wordpress/android/fluxc/persistence/SiteSqlUtils.java]: Ah, found a reference to that `TODO`: https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/58#issue-164517202

We should be able to close that issue too with this PR 🎉

**Comment by aforcier** [comment]
Looks good! I wrote some live site tests to check XML-RPC -> WP.com and WP.com -> XML-RPC and everything looks good: https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/5b044c9640713550cf60fec7caa442a4827e831b (PR forthcoming).

:shipit:

**Pull request merged**